common/quotas/global/doc.go

```diff
@@ -0,0 +1,158 @@
+
+"Aggregating" hosts will internally:
+  - receive data for key X from all "limiting" hosts
+  - compute request-weight per host+key in the cluster
+  - return the "limiting"-host's weighted portion of each ratelimit to allow
+
+The exact logic that the "aggregating" hosts use is intentionally hidden from the
+"limiting" hosts, so it can be changed without changing how they enforce limits.
+
+The weight-calculation algorithm will (eventually) be controlled by a dynamic config value,
+to allow shadowing and experimenting with different algorithms at runtime, or disabling
+the global logic altogether.
+
+See sub-packages for implementation details.
+
+# Data flow in the system
+
+The overall planned system's data-flow looks roughly like this:
+
+	┌───────┐
+	│users  │
+	└┬─┬─┬─┬┘
+	 a b c c
+	 │ │ │ │
+	┌▽─▽─▽─▽───┐
+	│frontend  │        (rejects based on previous or fallback limits, collects call counts)
+	│(limiting)│
+	└───┬──────┘
+	    │
+	 [a1,b1,c2]         (sends aggregated usage data to agg hosts, split by ring on keys)
+	    │
+	┌───▽─────────┐
+	│ring         │     (ringpop `Lookup(key)` to find agg host per key)
+	└┬─────┬─────┬┘
+	 │     │     │
+	 a1    b1    c2     (each subset of keys is sent to its owning agg host)
+	 │     │     │
+	 │     │     │  ┌─────── other "c" usage from other frontends, etc
+	┌▽───┐┌▽───┐┌▽──▽┐
+	│agg1││agg2││agg3│  (aggs compare frontend's usage per key with others, decide on RPS to allow)
+	└┬───┘└┬───┘└┬───┘
+	 │     │     │
+	a/2   b/4   c3/4    (aggs return weighted RPS to allow)
+	 │     │     │
+	┌▽─────▽─────▽┐
+	│frontend     │     (updates RPS, limits future calls, cycle continues)
+	└─────────────┘
+
+This package as a whole is only concerned with the high level request pattern in use:
+  - limiters enforce limits and collect data
+  - limiters asynchronously submit that data to all relevant aggregators, sharded by key
+  - aggregators decide per-limiter-and-key limits based on global data, and return it to limiters
+
+Currently:
+  - limiters are Frontends, but can trivially be any Cadence service as they only send outbound requests
+  - aggregators are arbitrarily in History service because it already has a ring implemented (Frontend does not)
+  - neither of these are fundamentally required, and they may change or be in multiple services later
+
+The actual contents of the requests and responses and how limits are decided is
+intentionally flexible and not defined here - see individual implementations / sub-packages.
+These will likely evolve as needs change, and will ideally be pluggable and
+dynamicconfig-selectable and shadow-able at runtime without changing this overall structure.
+
+# Planned use
+
+The first use of this is intentionally focused on addressing imbalanced usage
+of per-domain user/worker/ES ratelimits across different frontend hosts, e.g.
+due to unfair load balancing or per-datacenter affinity or other
+occasionally-changing-else-mostly-stable routing decisions.
```

Comment:
```suggestion
occasionally-changing-else-mostly-stable routing decisions.
This will essentially imply 3 keys per domain, and will look something like
`domain-user-request` / `domain-worker-request` / etc - they just have to be
unique per conceptual thing-with-a-configured-limit.
```